File: packages/python-sdk/strait/authoring/_agent.py
"""Agent definition DSL — wraps defineJob with agent conventions."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable

from strait.authoring._job import JobDefinition, JobOptions, define_job
from strait.authoring._run_context import RunContext


@dataclass
class AgentRunContext(RunContext):
    iteration: int = 0
    _accumulated_cost_microusd: int = 0
    _max_cost_microusd: float = float("inf")

    def accumulated_cost_microusd(self) -> int:
        return self._accumulated_cost_microusd

    def is_budget_exceeded(self) -> bool:
        return self._accumulated_cost_microusd >= self._max_cost_microusd


@dataclass
class AgentOptions:
    name: str = ""
    slug: str = ""
    endpoint_url: str = ""
    project_id: str = ""
    description: str = ""
    tags: dict[str, str] | None = None
    max_iterations: int | None = None
    max_cost_microusd: int | None = None
    auto_checkpoint: bool = True
    timeout_secs: int | None = None
    max_attempts: int | None = None
    retry_strategy: str = ""
    run: Callable[..., Any] | None = None
    on_success: Callable[..., Any] | None = None
    on_failure: Callable[..., Any] | None = None
    on_start: Callable[..., Any] | None = None


def define_agent(opts: AgentOptions) -> JobDefinition[Any]:
    max_cost = opts.max_cost_microusd if opts.max_cost_microusd is not None else float("inf")
    auto_checkpoint = opts.auto_checkpoint
    user_run = opts.run

    def agent_run(payload: Any, ctx: RunContext) -> Any:
        accumulated_cost = 0
        iteration = 0

        agent_ctx = AgentRunContext(
            run_id=ctx.run_id,
            attempt=ctx.attempt,
            logger=ctx.logger,
            checkpoint=None,
            report_progress=ctx.report_progress,
            heartbeat=ctx.heartbeat,
            report_usage=None,
            log_tool_call=ctx.log_tool_call,
            save_output=ctx.save_output,
            state=ctx.state,
            stream_chunk=ctx.stream_chunk,
            wait_for_event=ctx.wait_for_event,
            spawn=ctx.spawn,
            continue_run=ctx.continue_run,
            annotate=ctx.annotate,
            complete=ctx.complete,
            fail=ctx.fail,
            iteration=iteration,
            _accumulated_cost_microusd=0,
            _max_cost_microusd=max_cost,
        )

        original_report_usage = ctx.report_usage

        async def wrapped_report_usage(
            provider: str,
            model: str,
            prompt_tokens: int | None = None,
            completion_tokens: int | None = None,
            total_tokens: int | None = None,
            cost_microusd: int | None = None,
            **kwargs: Any,
        ) -> None:
            nonlocal accumulated_cost
            if cost_microusd is not None:
                accumulated_cost += cost_microusd
                agent_ctx._accumulated_cost_microusd = accumulated_cost
            if original_report_usage is not None:
                await original_report_usage(
                    provider=provider,
                    model=model,
                    prompt_tokens=prompt_tokens,
                    completion_tokens=completion_tokens,
                    total_tokens=total_tokens,
                    cost_microusd=cost_microusd,
                )

        if ctx.report_usage is not None:
            agent_ctx.report_usage = wrapped_report_usage

        original_checkpoint = ctx.checkpoint

        async def wrapped_checkpoint(state: dict[str, Any]) -> None:
            nonlocal iteration
            iteration += 1
            agent_ctx.iteration = iteration
            if auto_checkpoint and original_checkpoint is not None:
                await original_checkpoint(state)

        agent_ctx.checkpoint = wrapped_checkpoint

        if user_run is not None:
            return user_run(payload, agent_ctx)
        return None

    tags = dict(opts.tags or {})
    tags["strait.kind"] = "agent"

    return define_job(JobOptions(
        name=opts.name,
        slug=opts.slug,
        endpoint_url=opts.endpoint_url,
        project_id=opts.project_id,
        description=opts.description,
        tags=tags,
        timeout_secs=opts.timeout_secs or 600,
        max_attempts=opts.max_attempts or 5,
        retry_strategy=opts.retry_strategy or "exponential",
        run=agent_run,
        on_success=opts.on_success,
        on_failure=opts.on_failure,
        on_start=opts.on_start,
    ))

File: packages/ruby-sdk/lib/strait/authoring/agent.rb
# frozen_string_literal: true

module Strait
  module Authoring
    # Agent-specific run context with iteration and cost tracking.
    class AgentRunContext < RunContext
      attr_accessor :iteration

      def initialize(max_cost_microusd: Float::INFINITY, **kwargs)
        super(**kwargs)
        @iteration = 0
        @accumulated_cost_microusd = 0
        @max_cost_microusd = max_cost_microusd
      end

      def accumulated_cost_microusd
        @accumulated_cost_microusd
      end

      def add_cost(cost)
        @accumulated_cost_microusd += cost
      end

      def budget_exceeded?
        @accumulated_cost_microusd >= @max_cost_microusd
      end
    end

    # Options for defining an agent.
    AgentOptions = Struct.new(
      :name, :slug, :endpoint_url, :project_id, :description, :tags,
      :max_iterations, :max_cost_microusd, :auto_checkpoint, :timeout_secs,
      :max_attempts, :retry_strategy, :run, :on_success, :on_failure, :on_start,
      keyword_init: true
    ) do
      def initialize(**)
        super
        self.auto_checkpoint = true if auto_checkpoint.nil?
      end
    end

    # Creates a job definition with agent conventions.
    def self.define_agent(opts)
      max_cost = opts.max_cost_microusd || Float::INFINITY
      auto_checkpoint = opts.auto_checkpoint
      user_run = opts.run

      agent_run = ->(payload, ctx) {
        agent_ctx = AgentRunContext.new(
          run_id: ctx.run_id,
          attempt: ctx.attempt,
          report_progress: ctx.report_progress,
          heartbeat: ctx.heartbeat,
          log_tool_call: ctx.log_tool_call,
          save_output: ctx.save_output,
          state: ctx.state,
          stream_chunk: ctx.stream_chunk,
          wait_for_event: ctx.wait_for_event,
          spawn: ctx.spawn,
          continue_run: ctx.continue_run,
          annotate: ctx.annotate,
          complete: ctx.complete,
          fail: ctx.fail,
          max_cost_microusd: max_cost
        )

        # Wrap report_usage to track cost
        if ctx.report_usage
          original_report_usage = ctx.report_usage
          agent_ctx.report_usage = ->(**kwargs) {
            if kwargs[:cost_microusd]
              agent_ctx.add_cost(kwargs[:cost_microusd])
            end
            original_report_usage.call(**kwargs)
          }
        end

        # Wrap checkpoint to track iterations
        original_checkpoint = ctx.checkpoint
        agent_ctx.checkpoint = ->(state) {
          agent_ctx.iteration += 1
          original_checkpoint&.call(state) if auto_checkpoint
        }

        user_run&.call(payload, agent_ctx)
      }

      tags = (opts.tags || {}).merge("strait.kind" => "agent")

      job_opts = JobOptions.new(
        name: opts.name,
        slug: opts.slug,
        endpoint_url: opts.endpoint_url,
        project_id: opts.project_id,
        description: opts.description,
        tags: tags,
        timeout_secs: opts.timeout_secs || 600,
        max_attempts: opts.max_attempts || 5,
        retry_strategy: opts.retry_strategy || "exponential",
        run: agent_run,
        on_success: opts.on_success,
        on_failure: opts.on_failure,
        on_start: opts.on_start
      )

      JobDefinition.new(job_opts)
    end
  end
end
